tools/pubsys/src/repo/validate_repo/mod.rs

```diff
@@ -0,0 +1,193 @@
+    root_role_path: PathBuf,
+
+    #[structopt(default_value = "100", long)]
+    /// Randomly samples and retrieves specified percentage of targets
+    percent_target_files: u8,
+}
+
+/// Randomly samples specified percentage of listed targets in the TUF repo and tries to retrieve them.
+async fn retrieve_percentage_of_targets<T: 'static>(
+    repo: &Repository<'_, T>,
+    percentage: u8,
+) -> Result<(), Error>
+where
+    T: tough::Transport,
+    <T as tough::Transport>::Stream: std::marker::Send,
+{
+    let targets = &repo.targets().signed.targets;
+    let percentage = f32::from(percentage) / 100.0;
+    let num_to_retrieve = (targets.len() as f32 * percentage).ceil();
+    let mut rng = &mut rand::thread_rng();
+
+    let mut tasks = Vec::new();
+    for target in targets
+        .keys()
+        .collect::<Vec<&String>>()
+        .choose_multiple(&mut rng, num_to_retrieve as usize)
+        .cloned()
+    {
+        let target = target.to_string();
+        let mut reader = repo
+            .read_target(&target)
+            .with_context(|| repo_error::ReadTarget {
+                target: target.to_string(),
+            })?
+            .with_context(|| error::TargetMissing {
+                target: target.to_string(),
+            })?;
+        info!("Downloading target: {}", target);
+        tasks.push(tokio::spawn(async move {
+            // tough's `Read` implementation validates the target as it's being downloaded
+            io::copy(&mut reader, &mut io::sink()).context(error::TargetDownload {
+                target: target.to_string(),
+            })
+        }));
+    }
+    let results = join_all(tasks).await;
+    for result in results {
+        result.context(error::Join)??;
+    }
+
+    Ok(())
+}
+
+async fn validate_repo(
+    transport: &RepoTransport,
+    root_role_path: &PathBuf,
+    metadata_url: Url,
+    targets_url: &Url,
+    percent_target_files: u8,
+) -> Result<(), Error> {
+    // Create a temporary directory where the TUF client can store metadata
+    let workdir = tempdir().context(repo_error::TempDir)?;
+    let settings = Settings {
+        root: File::open(root_role_path).context(repo_error::File {
+            path: root_role_path,
+        })?,
+        datastore: workdir.path(),
+        metadata_base_url: metadata_url.as_str(),
+        targets_base_url: targets_url.as_str(),
+        limits: Limits::default(),
+        expiration_enforcement: ExpirationEnforcement::Safe,
+    };
+
+    // Load the repository
+    let repo = Repository::load(transport, settings).context(repo_error::RepoLoad {
+        metadata_base_url: metadata_url.clone(),
+    })?;
+    info!("Loaded TUF repo: {}", metadata_url);
+    // Try retrieving listed targets
+    info!(
+        "Attempting to download {}% of listed targets from {} for validation purposes",
+        percent_target_files, metadata_url
+    );
+    retrieve_percentage_of_targets(&repo, percent_target_files).await?;
```

Comment:
It's hard for me to reason about why we need this or what percentages other than 0% or 100% would make sense.

The advantage of the random subset is that we can converge on full validation over time without paying the cost during each run, but the drawback is that it's hard to know precisely when that will happen. It also doesn't help us put a time bound on execution since some of the targets are much larger than others (migration binary vs. root filesystem image).

What do you think about having a modal behavior, either "deep" (100%) or "shallow" (0%)? Then shallow checks could be scheduled more frequently and deep checks could run hourly or daily.

Reply:
Yeah, ideally the targets sampling would have worked more deterministically to cover all targets over some number of runs. But to implement something like that would be pretty complex and probably more trouble than it's worth.

I'll go ahead and change `--percent-target-files` to an optional flag `--validate-targets` to download all targets. By default we would just load the repository and make sure we can list the targets.

I think maybe a sensible addition would be to target files for specific versions of Bottlerocket (images, migrations) for validation but that is probably out of scope of this PR.

Reply:
I think we'll have to figure out a solution for validating targets without linearly increasing the time between metrics as we have more releases.  (Even percentage wouldn't do that.)  If we had a way of distinguishing new versus old targets (for us, the version numbers in the filenames?) then we could validate newer targets more frequently and do full runs as a batch job, perhaps.

Reply:
I will create an issue to track this once this PR is merged.

Reply:
https://github.com/bottlerocket-os/bottlerocket/issues/1172